core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala

```diff
@@ -0,0 +1,241 @@
+
+private[spark] class DriverLogger(conf: SparkConf)
+    extends Logging {
+
+  private val UPLOAD_CHUNK_SIZE = 1024 * 1024
+  private val UPLOAD_INTERVAL_IN_SECS = 5
+  private val DRIVER_LOG_FILE = "driver.log"
+  private val LOG_FILE_PERMISSIONS = new FsPermission(Integer.parseInt("770", 8).toShort)
+
+  private var localLogFile: String = FileUtils.getFile(
+    Utils.getLocalDir(conf), "driver_logs", DRIVER_LOG_FILE).getAbsolutePath()
+  private var hdfsLogFile: String = _
+  private var writer: Option[HdfsAsyncWriter] = None
+  private var hadoopConfiguration: Configuration = _
+  private var fileSystem: FileSystem = _
+
+  private val syncToDfs: Boolean = conf.get(DRIVER_LOG_SYNCTODFS)
+  private val syncToYarn: Boolean = conf.get(DRIVER_LOG_SYNCTOYARN)
+
+  addLogAppender()
+
+  private def addLogAppender(): Unit = {
+    val appenders = LogManager.getRootLogger().getAllAppenders()
+    val layout = if (appenders.hasMoreElements()) {
+      appenders.nextElement().asInstanceOf[Appender].getLayout()
+    } else {
+      new PatternLayout(conf.get(DRIVER_LOG_LAYOUT))
+    }
+    val fa = new Log4jFileAppender(layout, localLogFile)
+    fa.setName(DriverLogger.APPENDER_NAME)
+    LogManager.getRootLogger().addAppender(fa)
+    logInfo(s"Added a local log appender at: ${localLogFile}")
+  }
+
+  def startSync(hadoopConf: Configuration): Unit = {
+    try {
+      val appId = Utils.sanitizeDirName(conf.getAppId)
+      hdfsLogFile = {
+        val rootDir = conf.get(DRIVER_LOG_DFS_DIR.key,
+          "/tmp/driver_logs").split(",").head
+        FileUtils.getFile(rootDir, appId, DRIVER_LOG_FILE).getAbsolutePath()
+      }
+      hadoopConfiguration = hadoopConf
+      fileSystem = FileSystem.get(hadoopConf)
+
+      // Setup a writer which moves the local file to hdfs continuously
+      if (syncToDfs) {
+        writer = Some(new HdfsAsyncWriter())
+        logInfo(s"The local driver log file is being synced to spark dfs at: ${hdfsLogFile}")
+      }
+    } catch {
+      case e: Exception =>
+        logError(s"Could not sync driver logs to spark dfs", e)
+    }
+  }
+
+  def stop(): Unit = {
+    try {
+      writer.map(_.closeWriter())
+      if (syncToYarn) {
+        moveToYarnAppDir()
+      }
+    } catch {
+      case e: Exception =>
+        logError(s"Error in persisting driver logs", e)
+    } finally {
+      try {
+        JavaUtils.deleteRecursively(FileUtils.getFile(localLogFile).getParentFile())
+      } catch {
+        case e: Exception =>
+          logError(s"Error in deleting local driver log dir", e)
+      }
+    }
+  }
+
+  private def moveToYarnAppDir(): Unit = {
+    try {
+      val appId = Utils.sanitizeDirName(conf.getAppId)
+      val yarnConf = new YarnConfiguration(hadoopConfiguration)
+      val rootDir = yarnConf.get(YarnConfiguration.NM_REMOTE_APP_LOG_DIR)
+      if (rootDir != null && !rootDir.isEmpty()) {
+        var parentDir = if (UserGroupInformation.getCurrentUser() != null) {
```

Comment:
`val`

I don't know if `getCurrentUser` will ever be null. But if you're worried I'd just log something and do nothing here. The behavior of your code in that case does not look very appropriate (writing to "/yarn-root-dir/logs/blah") and would probably cause an exception anyway.